src/mlpack/methods/ann/init_rules/random_init.hpp

```diff
@@ -73,7 +73,7 @@ class RandomInitialization
                   const size_t cols,
                   const size_t slices)
   {
-    W = arma::Cube<eT>(rows, cols, slices);
+    W.set_size(rows, cols, slices);
```

Comment:
I am not sure how this will work with the memory pointer constructor. Anyway I suggest to replace `RandomInitialization` by `module-Weight().randu()`.

Reply:
I didn't think of that earlier. I have made the change as you suggested. Thanks for pointing it out :)
Do you think we should do the same thing for other init rules and perform test over each of them, to check whether it is possible to use other initializers in a similar way for weight initialization?

Reply:
> I didn't think of that earlier.

Perhaps
```c++
W.set_size(rows, cols, slices);
```
would work fine. I don't think Armadillo reallocates memory if it's not needed.

> Do you think we should do the same thing for other init rules

I'm not sure. I think using this function in order to initialize a matrix obtained by the memory pointer constructor could be dangerous. Suppose you pass an incorrect shape by mistake. Then the function will reallocate the matrix.

That is you don't need to pass the shape for any preallocated matrix. Perhaps it's better to introduce a special method for preallocated matrices:
```c++
void Initialize(arma::Cube<eT>& W)
```
@zoq Let's wait for Marcus'es response.

> and perform test over each of them

Currently you wrote about 4 tests with `RandomInitialization`. You can use different initializers in them. I don't think we need to increase the number of tests.

Reply:
Adding another method for preallocated matrices sounds like the safest option to me.